spring4/src/main/java/org/jdbi/v3/spring4/JdbiTransactionObject.java

```diff
@@ -0,0 +1,82 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.jdbi.v3.spring4;
+
+import org.springframework.transaction.CannotCreateTransactionException;
+import org.springframework.transaction.SavepointManager;
+import org.springframework.transaction.TransactionException;
+import org.springframework.transaction.TransactionSystemException;
+
+/**
+ * savepoint capable jdbi transaction wrapper
+ */
+public class JdbiTransactionObject implements SavepointManager {
+
+    private static final String JDBI_SAVEPOINT_BASENAME = "JDBI_SAVEPOINT_";
+
+    private int nextSavePointNumber = 0;
```

Comment:
should this be an `AtomicInteger`?

Reply:
hmm. on principal the transaction and any transaction related objects are afaik thread bound (at least in the local context of a spring application). that means no more than one thread will ever increment the number of save points.

I will look that up though and make it safe for general purposes if necessary